src/Analyzers/Core/Analyzers/ForEachCast/AbstractForEachCastDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,147 @@
+
+            // Don't bother checking conversions that are problematic for other reasons.  The user will already have a
+            // compiler error telling them the foreach is in error.
+            if (!conversion.Exists)
+                return;
+
+            // If the conversion was implicit, then everything is ok.  Implicit conversions are safe and do not throw at runtime.
+            if (conversion.IsImplicit)
+                return;
+
+            // An implicit legal conversion still shows up as explicit conversion in the object model.  But this is fine
+            // to keep as is since being an implicit-conversion means the API indicates it should always be safe to
+            // happen at runtime.
+            if (conversion.IsUserDefined && conversion.MethodSymbol is { Name: WellKnownMemberNames.ImplicitConversionName })
+                return;
+
+            if (collectionElementType is null)
+                return;
+
+            // We had a conversion that was explicit.  These are potentially unsafe as they can throw at runtime.
+            // Generally, we would like to notify the user about these.  However, we have different policies depending
+            // on if we think this is a legacy API or not.  Legacy APIs are those built before generics, and thus often
+            // have APIs that will just return `objects` and thus always need some sort of cast to get them to the right
+            // type.  A good example of that is S.T.RegularExpressions.CaptureCollection.  Users will almost always
+            // write this was `foreach (Capture capture in match.Captures)` and it would be annoying to force them to
+            // change this.
+            //
+            // What we do want to warn on are things like: `foreach (IUnrelatedInterface iface in stronglyTypedCollection)`.
+            //
+            // So, to detect if we're in a legacy situation, we look for iterations that are returning an object-type
+            // where the collection itself didn't implement `IEnumerable<T>` in some way.
+            if (option.Value == ForEachExplicitCastInSourcePreference.NonLegacy &&
+                IsLegacyAPI(ienumerableOfTType, collectionType, collectionElementType))
+            {
+                return;
+            }
+
+            // The user either always wants to write these casts explicitly, or they were calling a non-legacy API.
+            // report the issue so they can insert the appropriate cast.
+
+            // We can only fix this issue if the collection type implemented ienumerable.  Then we can add a .Cast
+            // call to it.
+            var isFixable = collectionType.AllInterfaces.Any(i => i.Equals(ienumerableType));
+
+            var options = semanticModel.Compilation.Options;
+            context.ReportDiagnostic(DiagnosticHelper.Create(
+                Descriptor,
+                node.GetFirstToken().GetLocation(),
+                Descriptor.GetEffectiveSeverity(options),
```

Comment:
I recall @mavasani mentioned that `option.Notification.Severity` should be used to support the `option = value:severity` syntax. Not sure if that applies here.

Reply:
wfm.